### PR TITLE
[FLINK-19773][runtime] Implement ExponentialDelayStrategy

### PR DESCRIPTION
**Pull request by mareksabo**
## What is the purpose of the change

*This PR add new exponential restart strategy as described in FLINK-19773.*

## Brief change log

Added restart strategy for exponential backoff.

## Verifying this change

- This change added unit tests, see PR.
- I manually verified that the restart strategy behaves as expected.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / no / **don't know**)
  - The runtime per-record code paths (performance sensitive): (yes / no / **don't know**)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (**yes, but only if the strategy is configured** / no / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / **docs** / **JavaDocs** / not documented)

### CODE REVIEW

**Review by tillrohrmann — APPROVED** [review]
Thanks for creating this PR @mareksabo. It looks very good. I tried the changes out and they work well. The only thing which you need to do is to rerun `mvn package -Dgenerate-config-docs -pl flink-docs,flink-core -DskipTests` in order to rebuild the configuration documentation. Once this is done and AZP passes, I will merge the PR.

File: flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/flip1/ExponentialDelayRestartBackoffTimeStrategyTest.java [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+
+		assertCorrectRandomRange(restartStrategy::getBackoffTime, 0L, 1L, 2L);
+
+		restartStrategy.notifyFailure(failure);
+		assertCorrectRandomRange(restartStrategy::getBackoffTime, 0L, 1L, 2L, 3L, 4L);
+
+		restartStrategy.notifyFailure(failure);
+		assertCorrectRandomRange(
+			restartStrategy::getBackoffTime,
+			0L, 1L, 2L, 3L, 4L, 5L, 6L, 7L, 8L);
```
[the remark this answers is not in the document]
Comment (mareksabo) [reply flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/failover/flip1/ExponentialDelayRestartBackoffTimeStrategyTest.java]: You are right, nice catch!

**Comment by mareksabo** [comment]
@flinkbot run azure

**Review by tillrohrmann — APPROVED** [review]
Thanks a lot for updating this PR @mareksabo. Merging this PR once AZP gives green light.

**Comment by mareksabo** [comment]
Thank you. Although I generated the documentation there was still some information missing in the docs -- I hope I just fixed that.

@flinkbot run azure

**Comment by mareksabo** [comment]
@flinkbot run azure

**Comment by tillrohrmann** [comment]
Really good work @mareksabo!

**Pull request closed without merging**